metric_tank/aggmetric.go

```diff
@@ -247,33 +253,117 @@ func (a *AggMetric) addAggregators(ts uint32, val float64) {
 	}
 }
 
-func (a *AggMetric) Persist(c *Chunk) {
-	log.Debug("AggMetric %s Persist(): starting to save %v", a.Key, c)
-	data := c.Series.Bytes()
-	chunkSizeAtSave.Value(int64(len(data)))
+func (a *AggMetric) persist(pos int) {
+	chunk := a.Chunks[pos]
+	chunk.Finish()
+	if *dryRun {
+		chunk.Saved = true
+		return
+	}
 
-	version := FormatStandardGoTsz
-	buf := new(bytes.Buffer)
-	err := binary.Write(buf, binary.LittleEndian, uint8(version))
-	if err != nil {
-		// TODO
+	// this will block when the write queue fills up.
+	log.Debug("sending chunk to write queue")
+	timer := time.NewTicker(2 * time.Second)
+WAIT:
+	for {
+		select {
+		case a.writeQueue <- chunk:
+			log.Debug("chunk in write queue: length: %d", len(a.writeQueue))
+			break WAIT
+		case <-timer.C:
+			log.Warn("%s:%d blocked pushing to writeQueue.", a.Key, chunk.T0)
+		}
 	}
-	_, err = buf.Write(data)
-	if err != nil {
-		// TODO
+	timer.Stop()
+
+	if !a.activeWrite {
+		log.Debug("starting persist goroutine.")
+		a.activeWrite = true
+		// asynchronously write data to cassandra.
+		go func() {
+			for {
+				select {
+				case c := <-a.writeQueue:
+					log.Debug("starting to save %s:%d %v", a.Key, c.T0, c)
+					data := c.Series.Bytes()
+					chunkSizeAtSave.Value(int64(len(data)))
+					version := FormatStandardGoTsz
+					buf := new(bytes.Buffer)
+					err := binary.Write(buf, binary.LittleEndian, uint8(version))
+					if err != nil {
+						// TODO
+					}
+					_, err = buf.Write(data)
+					if err != nil {
+						// TODO
+					}
+					success := false
+					attempts := 0
+					for !success {
+						err := InsertChunk(a.Key, c.T0, buf.Bytes(), *metricTTL)
+						if err == nil {
+							success = true
+							go func() {
+								a.Lock()
+								c.Saved = true
+								a.Unlock()
+							}()
+							log.Debug("save complete. %s:%d %v", a.Key, c.T0, c)
+							chunkSaveOk.Inc(1)
+						} else {
+							if (attempts % 20) == 0 {
+								log.Warn("failed to save metric to cassandra after %d attempts. %v, %s", attempts+1, c, err)
+							}
+							chunkSaveFail.Inc(1)
+							sleepTime := 100 * attempts
+							if sleepTime > 2000 {
+								sleepTime = 2000
+							}
+							time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+							attempts++
+						}
+					}
+				default:
+					log.Debug("waiting for lock: %s", a.Key)
+					a.Lock()
+					if len(a.writeQueue) == 0 {
```

Comment:
once we fall into this default case writeQueue should typically be 0,
but it may be >0 if persist() was called concurrently and just added something new. right? seems like a useful bit to add.